include/at_cmd_parser/at_params.h

```diff
@@ -57,8 +72,8 @@ union at_param_value {
 
 /** Parameter consisting of parameter type and value. */
 struct at_param {
-	enum at_param_type   type;
-	size_t               size;
+	enum at_param_type type;
```

Comment:
I still think this is a bit strange, I think the parser should be extended to allow for =,? characters in the command state so the string element would include those characters. Then helper function could figure out if it's an notification, set, read or test command. This would leave the at_parms code unchanged.

Reply:
I disagree - the information about the command type is resolved during the parsing process and should be returned by the parsing function. Hiding this information within the command string will result in:
1) a need to parse this information again,
2) and will make any string operations on the command string more difficult than they already are (like `strcmp`)

Reply:
Ok, then we have to agree to disagree:

1. No, you don't need parse the information again, you just accept more information into the string element when parsing a command. It's the same as you need to keep the %, # and + sign to see if it's an proprietary, custom or standard AT command. So what you do is to postpone the parsing of what it is. It simplifies your patch, doesn't change at_params module, keeps the parser code simpler and doesn't open up to add what ever special case into the core parser code.
2. No, because the command is always index 0, so you only need to treat this element specially.

Reply:
> It simplifies your patch, doesn't change at_params module, keeps the parser code simpler and doesn't open up to add what ever special case into the core parser code.

I think we should simplify the API usage, not the patch. I'm not sure what special case you refer to, but different command types covered in this PR are specified by Nordic AT command specification, I just follow the standard here.

> No, because the command is always index 0, so you only need to treat this element specially.

There's no difference if the command string is first on n-th parameter. Having to deal with special characters manually complicates the application code. Let's consider the standard AT+CFUN command as a sample. In case we append special character into the command, in or der to determine what action to take I need to either:
1) Provide 3 copies of the command string, `"AT+CFUN="`, `"AT+CFUN?"`, `"AT+CFUN=?"` and `strcpy` against each of them (for now let's ignore the fact that strings returned by the `at_cmd_parser` are not null-terminated),
2) or `strcpy` against `"AT+CFUN"` and then manually analyze following characters (keeping in mind that some SET commands are empty and have neither of these special characters). I call this parsing.

An alternative is to compare against  `"AT+CFUN"` string and then check the command type field, as proposed in this PR.

And @metc @evenl you should settle in which direction do you want this PR to go. On one hand you request a way more intrusive API change, than current proposal (@metc), on the other hand you don't want me to touch at_cmd_parser code at all (@evenl). Perhaps we should just give up on this patch, and have application to deal manually with AT commands. I don't see what's the point of having the `at_cmd_parser` module then.

Reply:
I think you are missunderstanding what I want:

> I think we should simplify the API usage, not the patch.

I think we get both from my proposal. The only change I ask for in your PR is that you move the command type parser (lines 137 through 149) out of the core parser and into a separate function, which is still a part of the at_cmd_parser API. For many users it is not necessary to know what command type it is as they are only parsing notifications, so this sound for me like extra information that doesn't belong in the at_params module context.

> There's no difference if the command string is first on n-th parameter.

Yes, for sure, but the parser is written in a way that will only detect commands at the first index, because that is how AT is specified. In cases with multiple commands, the string will be spitted up in multiple strings where the command always is index 0. 

> Provide 3 copies of the command string, "AT+CFUN=", "AT+CFUN?", "AT+CFUN=?" and strcpy against each of them (for now let's ignore the fact that strings returned by the at_cmd_parser are not null-terminated), or strcpy against "AT+CFUN" and then manually analyze following characters (keeping in mind that some SET commands are empty and have neither of these special characters). I call this parsing.

I think you are complicating the problem, the null-terminator is a decision that was made earlier and is the current implementation of the driver. But anyway in your example, strcmp could have been replaced by
strncmp("AT_CFUN", param, strlen("AT+CFUN");
of
res = strncmp("AT+CFUN", param, MIN(strlen("AT+CFUN"), parm_length); which will work both with and without NULL termination. 

> Having to deal with special characters manually complicates the application code.

That is not what I am trying to convey, I only say that I think the command type parsing should be a separate helper function in the at_cmd_parser API. 

> the other hand you don't want me to touch at_cmd_parser code at all (@evenl). 

I don't say that you should not touch the at_cmd_parser, I only say that you can avoid touching the at_params module that the at_cmd_parser depends on.

But me and Christopher will have a call today, and then we will get back to you.

Reply:
Me an Christopher just had a call, and we agreed that we want a separate function to parse the command type. So looking at the PR now, it seems like the only update you need to do is to move the `enum at_cmd_type {}` out of at_params.h and into at_cmd_parser.h and remove `enum at_param_type type;` from the at_parm struct.

Reply:
I've moved `enum at_cmd_type` into the `at_cmd_parser.h` header.

I think you are confused about `enum at_param_type type;` though - this patch has nothing to do with this field and I doubt we want to remove it.

Reply:
Aha, you are absolutely right. I was probably just confused because of the alignment change so it came up as a + line.